plugins/tableselection/plugin.js

```diff
@@ -538,57 +754,43 @@
-		// before inserting it into temporary container.
-		tmpContainer.setHtml( dataProcessor.toHtml( evt.data.dataValue ), {
-			fixForBody: false
-		} );
-		pastedTable = tmpContainer.findOne( 'table' );
-
-		if ( !selection.getRanges().length || !selection.isInTable() && !( boundarySelection = isBoundarySelection( selection ) ) ) {
+		// Do not customize paste process in following cases:
+		// No cells are selected.
+		if ( !selectedCells.length ||
+			// It's single range that does not fully contain table element and is not boundary, e.g. collapsed selection within
+			// cell, part of cell etc.
+			( selectedCells.length === 1 && !rangeContainsTableElement( selection.getRanges()[ 0 ] )  && !boundarySelection ) ||
+			// It's a boundary position but with no table pasted.
+			( boundarySelection && !pastedTable ) ) {
 			return;
 		}
 
-		selectedCells = getSelectedCells( selection );
+		selectedTable = selectedCells[ 0 ].getAscendant( 'table' );
+		tableSel = new TableSelection( getSelectedCells( selection, selectedTable ) );
 
-		if ( !selectedCells.length ) {
-			return;
+		function getLastArrayItem( arr ) {
+			return arr[ arr.length - 1 ];
 		}
 
-		selectedTable = selectedCells[ 0 ].getAscendant( 'table' );
-		selectedCells = getSelectedCells( selection, selectedTable );
-		firstCell = selectedCells[ 0 ];
-		firstRow = firstCell.getParent();
-		lastCell = selectedCells[ selectedCells.length - 1 ];
-		lastRow = lastCell.getParent();
-
 		// Schedule selecting appropriate table cells after pasting. It covers both table and not-table
 		// content (#520).
 		editor.once( 'afterPaste', function() {
-			var toSelect = cellToPaste ?
-				getCellsBetween( new CKEDITOR.dom.element( pastedTableMap[ 0 ][ 0 ] ), cellToPaste ) :
-				selectedCells;
+			var toSelect = pastedTableMap ?
+				getCellsBetween( new CKEDITOR.dom.element( pastedTableMap[ 0 ][ 0 ] ),
+					new CKEDITOR.dom.element( getLastArrayItem( getLastArrayItem( pastedTableMap ) ) ) ) :
+				tableSel.cells.all;
 
 			fakeSelectCells( editor, toSelect );
 		} );
 
-		// Handle mixed content (if the table is not the only child in the tmpContainer, we
-		// are probably dealing with mixed content). We handle also non-table content here.
-		// We just collapse selection to the first selected cell and pass non-table/mixed
-		// content to other paste handlers (#520).
-		if ( tmpContainer.getChildCount() > 1 || !pastedTable ) {
-			evt.data.dataValue = tmpContainer.getHtml();
-
-			selectCellContents( selectedCells[ 0 ] );
+		// In case of mixed content or non table content just insert it to a first cell, erasing the others.
+		if ( !pastedTable ) {
+			selectCellContents( tableSel.cells.first );
 
 			// Due to limitations of our undo manager, in case of mixed content
 			// cells must be emptied after pasting (#520).
 			editor.once( 'afterPaste', function() {
-				emptyCells( selectedCells.slice( 1 ), true );
+				emptyCells( tableSel.cells.all.slice( 1 ), true );
```

Comment:
Did it although emptyCells no longer features undo logic (it's been inlined in host code) as it's only used in that one single place.